File: dsa/4-queue/circular_queue.py
class CircularQueue():

    def __init__(self, size):
        self.size = size
        self.queue = [None] * size
        self.front = self.rear = -1

    def is_full(self):
        if (self.rear+1) % self.size == self.front:
            return True
        return False
    
    def is_empty(self):
        return True if self.rear==-1  else False
    
    # Insert an element into the circular queue
    def enqueue(self, data):

        if(self.is_full()):
            print("queue is full")
            return
        if (self.front == -1):
            self.front,self.rear = 0
            self.queue[self.rear] = data
        else:
            self.rear = (self.rear + 1) % self.size
            self.queue[self.rear] = data

    # Delete an element from the circular queue
    def dequeue(self):
        if (self.is_empty()):
            print("The circular queue is empty\n")
            return

        if (self.rear == self.front):
            temp = self.queue[self.front]
            self.front,self.rear = -1
            return temp
        else:
            temp = self.queue[self.front]
            self.front = (self.front + 1) % self.size
            return temp

    def printCQueue(self):
        if(self.is_empty()):
            print("No element in the circular queue")
            return

        if (self.rear >= self.front):
            for i in range(self.front, self.rear + 1):
                print(self.queue[i], end=" ")
            print()
        else:
            for i in range(self.front, self.size):
                print(self.queue[i], end=" ")
            for i in range(0, self.rear + 1):
                print(self.rear[i], end=" ")
            print()

if __name__=="__main__":
    q = CircularQueue(5)
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    q.enqueue(4)
    q.enqueue(5)
    q.printCQueue()

    q.dequeue()
    print("After removing an element from the queue")
    q.printCQueue()

File: dsa/04-queue/circularQueue.cpp
#include <iostream>
#define SIZE 5

using namespace std;

class CircularQueue
{
public:
    int front, rear, queue[SIZE];
    CircularQueue()
    {
        front = -1;
        rear = -1;
    }
    bool isEmpty();
    bool isFull();
    int enQueue(int data);
    int deQueue();
    void traverse();
};
bool CircularQueue::isEmpty()
{
    if (front == -1)
        return true;
    return false;
}

bool CircularQueue::isFull()
{
    if ((front == rear + 1) || (front == 0 && rear == SIZE + 1))
        return true;
    return false;
}
int CircularQueue::enQueue(int data)
{
    if (isFull())
        return 0;
    if (front == -1)
        front = 0;
    rear = (rear + 1) % SIZE;
    queue[rear] = data;
    return data;
}
int CircularQueue::deQueue()
{
    if (isEmpty())
        return 0;

    int data = queue[front];
    if (front == rear)
    {
        front = 1;
        rear = -1;
    }
    else
    {
        queue[front] = 0;
        front = (front + 1) % SIZE;
    }
    return data;
}

void CircularQueue::traverse()
{
    cout << "front: " << front << endl;
    for (int i = 0; i < SIZE; i++)
        cout << queue[i] << " ";
}
int main()
{
    CircularQueue q;
    q.enQueue(1);
    q.enQueue(2);
    q.enQueue(3);
    q.enQueue(4);
    q.enQueue(5);

    q.traverse();
    q.deQueue();
    q.traverse();
    q.enQueue(100);
    q.traverse();

    return 0;
}